helix-core/src/doc_formatter.rs

```diff
@@ -0,0 +1,384 @@
+            }
+        }
+    }
+
+    fn advance_grapheme(&mut self, col: usize) -> Option<FormattedGrapheme<'t>> {
+        let (grapheme, source) =
+            if let Some((grapheme, highlight)) = self.next_inline_annotation_grapheme() {
+                (grapheme.into(), GraphemeSource::VirtualText { highlight })
+            } else if let Some(grapheme) = self.graphemes.next() {
+                self.virtual_lines += self.annotations.annotation_lines_at(self.char_pos);
+                let codepoints = grapheme.len_chars() as u32;
+
+                let overlay = self.annotations.overlay_at(self.char_pos);
+                let grapheme = match overlay {
+                    Some((overlay, _)) => overlay.grapheme.as_str().into(),
+                    None => Cow::from(grapheme).into(),
+                };
+
+                self.char_pos += codepoints as usize;
+                (grapheme, GraphemeSource::Document { codepoints })
+            } else {
+                if self.exhausted {
+                    return None;
+                }
+                self.exhausted = true;
+                // EOF grapheme is required for rendering
+                // and correct position computations
+                return Some(FormattedGrapheme {
+                    grapheme: Grapheme::Other { g: " ".into() },
+                    source: GraphemeSource::Document { codepoints: 0 },
+                });
+            };
+
+        let grapheme = FormattedGrapheme::new(grapheme, col, self.text_fmt.tab_width, source);
+
+        Some(grapheme)
+    }
+
+    /// Move a word to the next visual line
+    fn wrap_word(&mut self, virtual_lines_before_word: usize) -> usize {
+        // softwrap this word to the next line
+        let indent_carry_over = if let Some(indent) = self.indent_level {
+            if indent as u16 <= self.text_fmt.max_indent_retain {
+                indent as u16
+            } else {
+                0
+            }
+        } else {
+            // ensure the indent stays 0
+            self.indent_level = Some(0);
+            0
+        };
+
+        self.visual_pos.col = indent_carry_over as usize;
+        self.virtual_lines -= virtual_lines_before_word;
+        self.visual_pos.row += 1 + virtual_lines_before_word;
+        let mut i = 0;
+        let mut word_width = 0;
+        let wrap_indicator = UnicodeSegmentation::graphemes(&*self.text_fmt.wrap_indicator, true)
+            .map(|g| {
+                i += 1;
+                let grapheme = FormattedGrapheme::new(
+                    g.into(),
+                    self.visual_pos.col + word_width,
+                    self.text_fmt.tab_width,
+                    GraphemeSource::VirtualText {
+                        highlight: self.text_fmt.wrap_indicator_highlight,
+                    },
+                );
+                word_width += grapheme.width();
+                grapheme
+            });
+        self.word_buf.splice(0..0, wrap_indicator);
```

Comment:
Based on the usage would a `VecDeque` work better here? Or a `String` since it supports `insert_str`

Reply:
A `VedDeque` could work but we append probably 95% of the time (well really `(width) / (wrap_indicator_width + width)` which is 97% for a width of 80 and default settings, usually width is larger). A `VecDeque` has quite a bit of overhead compared to a `Vec` so I don't think this single case of prepending justifies that.

The `word_buf` is usually just a couple graphemes long so the cost of copying that backwards here is pretty trivial.

A `String` is also a `Vec` internally so `insert_str` endsup being pretty equivalent to the above code. However I store graphemes pre-segmented so we don't have to run graphemes segementation multiple times.

Reply:
I assumed `insert_str`/`insert_bytes` is more efficient than splice https://stackoverflow.com/questions/47037573/how-to-prepend-a-slice-to-a-vec

Reply:
Ah that is what you mean. For an exact size iterator I would be surprised if a splice doesn't get optimized down to the same thing. Altough you are right that in this case `UnicodeSegementation::graphemes` is not an exact size iterator. I could optimize this by pre-segmenting the wrap_indicator (so we store a `Vec` of graphemes instead of a string), then we would have a size hint and can prepend the vec with two trivial memcopys.

I considered doing that when I initially implemented this but ended up just going with the splice because I didn't think the performance would matter for what is (usually) an iterator with 1 or 2 elements being prepended to a vec with atmost 20 elements (default settings). So I choose simplicity/readability over all-out performance. I am certain that a `VecDeque` and `String` would perform worse overall (repeated unicode segmentation or just overhead of a `VecDeque`) as line wrapping is pretty rare compared to just normal pushes.

If you think it's worth further complicating the code here to optimize this case, I could certainly cache the segmentation of the `wrap-indicator` when constructing the `DocFormatter`.

Reply:
Seems ok to leave this out for now